### PR TITLE
create unit tests for processors and services

### PR DESCRIPTION
**Pull request by yoursweet667**

### CODE REVIEW

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/LeaveGameProcessorTest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import ru.yoursweet667.uno.service.model.event.LeaveGameEvent;
+
+import java.util.HashMap;
+import java.util.Map;
+import static org.assertj.core.api.Assertions.assertThat;
+
+
+public class LeaveGameProcessorTest {
+
+    LeaveGameProcessor leaveGameProcessor = new LeaveGameProcessor();
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/LeaveGameProcessorTest.java:15]: can be private final. Same comment for other tests.

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/LeaveGameProcessorTest.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        players.put(player.getPlayerId(), player);
+        LeaveGameEvent event = new LeaveGameEvent(123, null, player);
+        Game game = new Game(null, players, null, null,
+                null, null);
+
+        //When
+        leaveGameProcessor.doProcess(event, game);
+
+        //Then
+        assertThat(game.getPlayers()).doesNotContainValue(player);
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/LeaveGameProcessorTest.java:32]: Instead, check that players map doesn't contain the playerId key. If there's no key - there's no value.

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/PlayCardProcessorTest.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        Player player = new Player("sourcePlayerId", null, cards);
+        PlayCardEvent event = new PlayCardEvent(123, "sourcePlayerId", null, card);
+        Game game = new Game(null, Map.of(player.getPlayerId(), player),
+                null, null, cardsInGame, null);
+
+        //When
+        playCardProcessor.doProcess(event, game);
+
+        //Then
+        assertThat(game.getCardsInTheGame()).contains(card);
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/PlayCardProcessorTest.java:35]: check that this is the last card in the game

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/SkipTurnProcessorTest.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+
+public class SkipTurnProcessorTest {
+
+    SkipTurnProcessor skipTurnProcessor = new SkipTurnProcessor();
+
+    @Test
+    void doProcess_setRightNextPlayer() {
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/SkipTurnProcessorTest.java:18]: 'Correct' word suits here better than 'Right'

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/TakeCardsProcessorTest.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        //When
+        takeCardsProcessor.doProcess(event, game);
+
+        //Then
+        assertThat(event.getPlayer().getCards()).contains(card);
+
+    }
+
+    @Test
+    void doProcess_anyGameState_takeCards() {
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/TakeCardsProcessorTest.java:39]: any game state?

File: UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+
+    @Test
+    void createGame_createsGame() {
+
+        //Given+When
+        Game game = gameService.createGame();
+
+        //Then
+        assertThat(game).isNotNull();
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java:42]: Check that games properties are initialized correctly. Check that there are cards, etc.

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/LeaveGameProcessorTest.java (hunk 1 above)
Comment (yoursweet667) [reply UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/LeaveGameProcessorTest.java:15]: done

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/LeaveGameProcessorTest.java (hunk 2 above)
Comment (yoursweet667) [reply UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/LeaveGameProcessorTest.java:32]: changes has been made

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/PlayCardProcessorTest.java (hunk 3 above)
Comment (yoursweet667) [reply UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/PlayCardProcessorTest.java:35]: made a check

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/SkipTurnProcessorTest.java (hunk 4 above)
Comment (yoursweet667) [reply UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/SkipTurnProcessorTest.java:18]: changed

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/TakeCardsProcessorTest.java (hunk 5 above)
Comment (yoursweet667) [reply UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/TakeCardsProcessorTest.java:39]: any allowed

File: UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java (hunk 6 above)
Comment (yoursweet667) [reply UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java:42]: Done

File: UNO/src/test/java/ru/yoursweet667/uno/dataaccess/game/InMemoryGameStorageTest.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    @Test
+    void getGame_returnGame() {
+
+        //Give
+        Game game = new Game(GAME_ID, null, null,
+                null, null, null);
+
+        Mockito.when(inMemoryGameStorage.getGame(GAME_ID)).thenReturn(Optional.of(game));
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/dataaccess/game/InMemoryGameStorageTest.java:82]: You're supposed to test `inMemoryGameStorage.getGame(GAME_ID)` here. But you're mocking its behavior, which means you're not testing `getGame`, but you're testing a mock.

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/EventServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        //When
+        int fromEventId = 0;
+        List<Event> events = eventService.getEvents(GAME_ID, PLAYER_ID, fromEventId);
+
+        //Then
+        assertThat(fromEventId).isLessThanOrEqualTo(events.size());
+    }
+
+    @Test
+    void getEvents_gameExistChangeEventParams_getNeededEvents() {
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/event/EventServiceImplTest.java:138]: This name is confusing. How about `getEvents_eventsContainOthePlayersSensitiveData_returnedEventsHaveHiddenDetails`

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/EndGameProcessorTest.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+import ru.yoursweet667.uno.service.model.GameState;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class EndGameProcessorTest {
+
+    private static final EndGameProcessor endGameProcessor = new EndGameProcessor();
+
+    @Test
+    void doProcess_anyGameState_changeGameState() {
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/EndGameProcessorTest.java:14]: Does `_anyGameState_` actually matter for this test? If not - just remove this.

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/EndTurnProcessorTest.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        endTurnProcessor.doProcess(event, game, resultEventConsumer);
+
+        //Then
+        assertThat(game.getGameState()).isEqualTo(GameState.END_TURN);
+        assertThat(game.getNextPlayer().get().getPlayerId()).isEqualTo("playerId");
+        assertThat(resultEventConsumer.event.getType()).isEqualTo(EventType.START_TURN);
+        assertThat(resultEventConsumer.gameId).isEqualTo(game.getGameId());
+    }
+
+    private static class ValueHoldingBiConsumer implements BiConsumer<String, Event> {
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/EndTurnProcessorTest.java:50]: Would be better to remove `ValueHoldingBiConsumer` from tests and use mock + mockito.verify instead, just like you've done for one of the test classes.

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/PlayCardProcessorTest.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+
+public class PlayCardProcessorTest {
+
+    private final PlayCardProcessor playCardProcessor = new PlayCardProcessor();
+
+    @Test
+    void doProcess_addCardInGameAndDeleteInPlayer() {
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/PlayCardProcessorTest.java:19]: _addCardToGameAndRemoveItFromPlayer

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/TakeCardsProcessorTest.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        //When
+        takeCardsProcessor.doProcess(event, game, null);
+
+        //Then
+        assertThat(event.getPlayer().getCards()).contains(card);
+
+    }
+
+    @Test
+    void doProcess_anyAllowedGameState_takeCards() {
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/TakeCardsProcessorTest.java:39]: You say `_anyAllowedGameState_`, but in the tests code I see `GameState.START_TURN`

File: UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+    void createGame_createsGame() {
+
+        //Given+When
+        Game game = gameServiceImpl.createGame();
+
+        //Then
+        assertThat(game).isNotNull();
+        assertThat(game.getGameId()).isNotNull();
+        assertThat(game.getPlayers()).isEmpty();
+        assertThat(game.getGameState()).isNotNull();
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java:45]: Should it be ANY not null state or a particular one?

File: UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+        assertThat(game.getDeck()).isNotEmpty();
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java:46]: Check that at least it has a correct number of cards. This test right now will pass even if `createGame()` will create a deck of just 1 card.

File: UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+        assertThat(game.getCardsInTheGame()).isEmpty();
+        assertThat(game.getEvents()).isEmpty();
+
+        Mockito.verify(storage).createGame(game);
+
+
+    }
+
+    @Test
+    void addPlayerToGame_addPlayerToLocalGame() {
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java:56]: I don't understand what `Local` means here.

File: UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+        Player player = gameServiceImpl.addPlayerToGame(GAME_ID, "playerName");
+
+        //Then
+        assertThat(game.getPlayers()).containsKey(player.getPlayerId());
+        assertThat(game.getPlayers().get(player.getPlayerId())).isEqualTo(player);
+
+    }
+
+    @Test
+    void addPlayerToGame_gameNotFound_exception() {
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java:74]: I don't see anything related to exception in this test's code.

File: UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+    void removePlayerFromGame_removePlayer() {
+
+        //Given
+        Game game = new Game(GAME_ID, new HashMap<>(), null,
+                null, null, null);
+
+        Mockito.when(storage.getGame(GAME_ID)).thenReturn(Optional.of(game));
+
+        //When
+        Player player = gameServiceImpl.addPlayerToGame(GAME_ID, "playerName");
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java:100]: why are you calling `addPlayerToGame` in `removePlayerFromGame_removePlayer` test?

File: UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+
+        //Then
+        assertThat(storage.getGame(GAME_ID)).isNotNull();
+
+    }
+
+    @Test
+    void removePlayerFromGame_gameNotFound_exception() {
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java:108]: don't see exceptions in the code of this test

File: UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+        Game game = new Game(GAME_ID, null, null,
+                null, null, null);
+
+        Mockito.when(storage.getGame(GAME_ID)).thenReturn(Optional.of(game));
+
+        //When
+        Optional<Game> gameFromStorage = gameServiceImpl.getGame(GAME_ID);
+
+        //Then
+        assertThat(gameFromStorage.get().getGameId()).isEqualTo(game.getGameId());
```
Comment (phreakiboi) [inline UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java:140]: Don't check that IDs are equal, check that games are equal

**Review by phreakiboi — CHANGES_REQUESTED** [review]

File: UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java (hunk 19 above)
Comment (yoursweet667) [reply UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java:140]: Fixed it

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/EndGameProcessorTest.java (hunk 9 above)
Comment (yoursweet667) [reply UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/EndGameProcessorTest.java:14]: ok

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/PlayCardProcessorTest.java (hunk 11 above)
Comment (yoursweet667) [reply UNO/src/test/java/ru/yoursweet667/uno/service/event/processor/PlayCardProcessorTest.java:19]: Changed that

File: UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java (hunk 15 above)
Comment (yoursweet667) [reply UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java:56]: actually nothing, deleted it

File: UNO/src/test/java/ru/yoursweet667/uno/service/event/EventServiceImplTest.java (hunk 8 above)
Comment (yoursweet667) [reply UNO/src/test/java/ru/yoursweet667/uno/service/event/EventServiceImplTest.java:138]: renamed it

File: UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java (hunk 18 above)
Comment (yoursweet667) [reply UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java:108]: this test was written incorrectly, now it is fixed

File: UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java (hunk 13 above)
Comment (yoursweet667) [reply UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java:45]: it must be GAME_CREATED

File: UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java (hunk 14 above)
Comment (yoursweet667) [reply UNO/src/test/java/ru/yoursweet667/uno/service/game/GameServiceImplTest.java:46]: fixed it

File: UNO/src/test/java/ru/yoursweet667/uno/dataaccess/game/InMemoryGameStorageTest.java (hunk 7 above)
Comment (yoursweet667) [reply UNO/src/test/java/ru/yoursweet667/uno/dataaccess/game/InMemoryGameStorageTest.java:82]: Fixed it

**Pull request closed without merging**